base/compiler/optimize.jl

```diff
@@ -176,25 +176,21 @@ _topmod(sv::OptimizationState) = _topmod(sv.mod)
 
 function isinlineable(m::Method, me::OptimizationState, params::OptimizationParams, union_penalties::Bool, bonus::Int=0)
     # compute the cost (size) of inlining this code
-    inlineable = false
     cost_threshold = params.inline_cost_threshold
     if m.module === _topmod(m.module)
         # a few functions get special treatment
         name = m.name
         sig = m.sig
         if ((name === :+ || name === :* || name === :min || name === :max) &&
             isa(sig,DataType) &&
-            sig == Tuple{sig.parameters[1],Any,Any,Any,Vararg{Any}})
-            inlineable = true
+            sig === Tuple{sig.parameters[1],Any,Any,Any,Vararg{Any}})
+            return true
```

Comment:
Why aren't we using
```diff
diff --git a/base/operators.jl b/base/operators.jl
index 82b393e0a9..dcc0c3704a 100644
--- a/base/operators.jl
+++ b/base/operators.jl
@@ -636,7 +636,7 @@ for op in (:+, :*, :&, :|, :xor, :min, :max, :kron)
         # note: these definitions must not cause a dispatch loop when +(a,b) is
         # not defined, and must only try to call 2-argument definitions, so
         # that defining +(a,b) is sufficient for full functionality.
-        ($op)(a, b, c, xs...) = afoldl($op, ($op)(($op)(a,b),c), xs...)
+        @inline ($op)(a, b, c, xs...) = afoldl($op, ($op)(($op)(a,b),c), xs...)
         # a further concern is that it's easy for a type like (Int,Int...)
         # to match many definitions, so we need to keep the number of
         # definitions down to avoid losing type information.
```
since 7bdf9c664dd539899afba6a23e03a288720a07fb / f8d229499c7bc6eced2d13cdd211767fe46b68cb (#11274 / #13350) / 86ffe49b95f642c15221050918844fe9f545551f (#5011)

Reply:
The only difference seems to be that the definition-site annotation is only respected when the call signature is concrete ?
https://github.com/JuliaLang/julia/blob/d6f59fa1826eb2db1ac1980af71cb6fcc4c7a978/base/compiler/optimize.jl#L296-L297

Well, I don't think it makes much difference in reality (we can't get best performance when there is type instability anyway), and so I'm inclined to simply switch to the annotation.